ot/bregman.py

```diff
@@ -1787,3 +1789,287 @@ def empirical_sinkhorn_divergence(X_s, X_t, reg, a=None, b=None, metric='sqeucli
+
+        cpt = 1
+        while (cpt < 5):  # 5 iterations
+            K_IJ_v = K_IJ.T @ u0 + cst_v
+            v0 = b_J / (kappa * K_IJ_v)
+            KIJ_u = K_IJ @ v0 + cst_u
+            u0 = (kappa * a_I) / KIJ_u
+            cpt += 1
+
+        u0 = projection(u0, epsilon / kappa)
+        v0 = projection(v0, epsilon * kappa)
+
+    else:
+        u0 = u0
+        v0 = v0
+
+    def restricted_sinkhorn(usc, vsc, max_iter=5):
+        """
+        Restricted Sinkhorn Algorithm as a warm-start initialized point for L-BFGS-B (see Algorithm 2 in [1]).
+        """
+        cpt = 1
+        while (cpt < max_iter):
+            K_IJ_v = K_IJ.T @ usc + cst_v
+            vsc = b_J / (kappa * K_IJ_v)
+            KIJ_u = K_IJ @ vsc + cst_u
+            usc = (kappa * a_I) / KIJ_u
+            cpt += 1
+
+        usc = projection(usc, epsilon / kappa)
+        vsc = projection(vsc, epsilon * kappa)
+
+        return usc, vsc
+
+    def screened_obj(usc, vsc):
+        part_IJ = usc @ K_IJ @ vsc - kappa * a_I @ np.log(usc) - (1. / kappa) * b_J @ np.log(vsc)
+        part_IJc = usc @ vec_eps_IJc
+        part_IcJ = vec_eps_IcJ @ vsc
+        psi_epsilon = part_IJ + part_IJc + part_IcJ
+        return psi_epsilon
+
+    def screened_grad(usc, vsc):
+        # gradients of Psi_epsilon w.r.t u and v
+        grad_u = K_IJ @ vsc + vec_eps_IJc - kappa * a_I / usc
+        grad_v = K_IJ.T @ usc + vec_eps_IcJ - (1. / kappa) * b_J / vsc
+        return grad_u, grad_v
+
+    def bfgspost(theta):
+
+        u = theta[:ns_budget]
+        v = theta[ns_budget:]
+        # objective
+        f = screened_obj(u, v)
+        # gradient
+        g_u, g_v = screened_grad(u, v)
+        g = np.hstack([g_u, g_v])
+        return f, g
+
+    #----------------------------------------------------------------------------------------------------------------#
+    #                                           Step 2: L-BFGS-B solver                                              #
+    #----------------------------------------------------------------------------------------------------------------#
+
+    u0, v0 = restricted_sinkhorn(u0, v0)
+    theta0 = np.hstack([u0, v0])
+    maxiter = 10000  # max number of iterations
+    maxfun = 10000  # max  number of function evaluations
+    pgtol = 1e-09  # final objective function accuracy
+
+    bounds = bounds_u + bounds_v  # constraint bounds
+    obj = lambda theta: bfgspost(theta)
+
+    theta, _, _ = fmin_l_bfgs_b(func=obj,
+                                x0=theta0,
+                                bounds=bounds,
+                                maxfun=maxfun,
+                                pgtol=pgtol,
+                                maxiter=maxiter)
+
+    usc = theta[:ns_budget]
+    vsc = theta[ns_budget:]
+
+    usc_full = np.full(ns, epsilon / kappa)
+    vsc_full = np.full(nt, epsilon * kappa)
+    usc_full[I] = usc
+    vsc_full[J] = vsc
+
+    Gsc = usc_full.reshape((-1, 1)) * K * vsc_full.reshape((1, -1))
```

Comment:
Pprefer the numpy slicing instead of reshape: 
```python
gamma = u[:, None] * K * v[None, :]
return gamma
```
This allows to "see" the formula gamma_ij = u_i K_ij v_j. 

This function should also have an argument `log` and return the transport and a dictionary `log` containing the dual variables (see the other sinkhorn functions).

Reply:
agreed with both comments